Doc/regions.md

```diff
@@ -0,0 +1,567 @@
+```python
+>>> locals_example2()
+{'x': 1}
+```
+
+The `locals()` function also captures the local variables in the enclosing scope.
+```python
+def locals_example3():
+  x = 1
+  y = 2
+  def inner():
+    t = x
+    return locals()
+  return inner()
+```
+This returns a dictionary object that contains the local variables in the enclosing scope that have been mentioned.
+```python
+>>> locals_example3()
+{'t': 1, 'x': 1}
+```
+The variable `x` has been captured by the `locals()` function,
+but `y` has not been mentioned in the inner function, and thus is not captured.
+
+The capture is also nested. 
+```python
+def locals_example4():
+  x = 1
+  y = 2
+  def inner():
+    def inner_inner():
+      t = x
+    return locals()
+  return inner()
+```
+Here the dictionary returned `locals` contains the function object `inner_inner` and `x`.
+Although, `inner` does not directly use `x`, it is captured by `inner_inner` and thus is included in the dictionary.
+The `t` used in `inner_inner` is not captured as it is in an inner scope.
+```python
+>>> locals_example4()
+{'inner_inner': <function locals_example4.<locals>.inner.<locals>.inner_inner at 0x7ea265d3add0>, 'x': 1}
+```
+
+The `locals()` function does not store the `cell` for closures in the dictionary.
+```python
+def locals_example5():
+  x = 1
+  def inc():
+    nonlocal x
+    x += 1
+    return x
+  return locals()
+```
+If we access the dictionary, the function object `inc` is stored, but the `cell` object is not.
+The original value of the variable `x` is stored in the dictionary.
+```python
+>>> d = locals_example5()
+>>> d["x"]
+1
+>>> d["inc"]()
+2
+>>> d["inc"]()
+3
+>>> d["x"]
+1
+```
+
+
+### Frame objects
+
+The `locals()` call provides a snapshot of the local variables in the current scope.
+However, it does not provide access to the frame object itself.
+The frame object is a Python object that represents the current execution context.
+
+The primary purpose is for debugging.
+The frame object is often used to be able to give a more comprehensive error/log message.
+This can involve potentially stack traces, or the local variables at the time of the error.
+
+The frame object contains a reference to the code object that is being executed.
+It also contains a reference to the global and local variables.
+Python programs can use this object to access the local and global variables.
+This object is updated in sync with the underlying variables:
+
+```python
+def frame_example1():
+  x = 1
+  l = sys._getframe()
+  x = 2
+  return l.f_locals
+```
+This behaves differently to `locals` and captures the update to `x`.
+```python
+>>> f = frame_example1()
+>>> f
+{'x': 2, 'l': <frame at 0x7ea265e76f60, file '<stdin>', line 5, code frame_example1>}
+```
+Interesting, it also contains `l` which is the frame object itself.
+This is not the case for `locals()`, which does not capture the dictionary it creates.
```

Comment:
The distinction is because you assigned it to `l`, not because of the difference in behavior between `locals()` and `sys._getframe()`.

Reply:
2a2b08ac2ecd983ea54607b24237fc69323d4d64